prow/tide/search.go

```diff
@@ -27,121 +27,50 @@ import (
-		}).Debugf("Query returned %d PRs and cost %d point(s). %d remaining.", len(ret), totalCost, remaining)
-		return ret, totalMatches, nil
-	}
-}
-
-func (q searchExecutor) search() ([]PullRequest, error) {
-	prs, _, err := q.searchRange(time.Time{}, time.Now())
-	return prs, err
-}
-
-func (q searchExecutor) searchSince(t time.Time) ([]PullRequest, error) {
-	prs, _, err := q.searchRange(t, time.Now())
-	return prs, err
-}
-
-func (q searchExecutor) searchRange(start, end time.Time) ([]PullRequest, int, error) {
-	// Adjust times to be after GitHub was founded to avoid querying empty time
-	// ranges.
+func search(query querier, log *logrus.Entry, q string, start, end time.Time) ([]PullRequest, error) {
 	if start.Before(github.FoundingYear) {
 		start = github.FoundingYear
 	}
 	if end.Before(github.FoundingYear) {
 		end = github.FoundingYear
 	}
-
-	prs, count, err := q(start, end)
-	if err != nil {
-		return nil, 0, err
-	}
-
-	if count <= 1000 {
-		// The search returned all the results for the query.
-		return prs, len(prs), nil
-	}
-	// The query returned too many results, we need to partition it.
-	prs, err = q.partitionSearchRange(start, end, count)
-	return prs, len(prs), err
-}
-
-func (q searchExecutor) partitionSearchRange(start, end time.Time, count int) ([]PullRequest, error) {
-	partition := partitionTime(start, end, count, 900)
-	// Search right side...
-	rPRs, rCount, err := q.searchRange(partition, end)
-	if err != nil {
-		return nil, err
-	}
-
-	// Search left side...
-	// For the left side we can deduce the count in advance.
-	lCount := count - rCount
-	// If the count is too large we can skip the initial search and go straight to
-	// partitioning to save an API token.
-	var lPRs []PullRequest
-	if lCount <= 1000 {
-		lPRs, _, err = q.searchRange(start, partition)
-	} else {
-		lPRs, err = q.partitionSearchRange(start, partition, lCount)
-	}
-	if err != nil {
-		return nil, err
+	log = log.WithFields(logrus.Fields{
+		"query": q,
+		"start": start.String(),
+		"end":   end.String(),
+	})
+	requestStart := time.Now()
+	datedQuery := fmt.Sprintf("%s %s", q, dateToken(start, end))
+	var cursor *githubql.String
+	vars := map[string]interface{}{
+		"query":        githubql.String(datedQuery),
+		"searchCursor": cursor,
 	}
-
-	return append(lPRs, rPRs...), nil
-}
-
-func partitionTime(start, end time.Time, count, goalSize int) time.Time {
-	duration := end.Sub(start)
-	if count < goalSize*2 {
-		// Choose the midpoint.
-		return start.Add(duration / 2)
+	var totalCost, remaining int
+	var ret []PullRequest
+	var sq searchQuery
+	for {
+		log.Debug("Sending query")
+		if err := query(context.Background(), &sq, vars); err != nil {
+			return ret, err
+		}
+		totalCost += int(sq.RateLimit.Cost)
+		remaining = int(sq.RateLimit.Remaining)
+		for _, n := range sq.Search.Nodes {
+			ret = append(ret, n.PullRequest)
+		}
+		if !sq.Search.PageInfo.HasNextPage {
+			break
+		}
+		cursor = &sq.Search.PageInfo.EndCursor
```

Comment:
This won't actually update the value of `vars["searchCursor"]` meaning that pagination won't work.  It would be nice to at least have a test case for the happy case so that we can catch bugs like this.

Reply:
Done

I would like to first validate this refactor works correctly before writing unit tests for it (aka we won't throw it away in an hour).